ktor-client/ktor-client-core/common/src/io/ktor/client/plugins/sse/DefaultClientSSESession.kt

```diff
@@ -4,36 +4,121 @@
+    private val maxRetries = content.maxRetries
 
-    private val _incoming = channelFlow {
-        while (true) {
-            val event = input.parseEvent() ?: break
+    private val initialRequest = content.initialRequest
+
+    private val clientForReconnection = initialRequest.attributes[SSEClientForReconnectionAttr]
+
+    public constructor(
+        content: SSEClientContent,
+        input: ByteReadChannel
+    ) : this(content, input, content.callContext + Job() + CoroutineName("DefaultClientSSESession"))
+
+    private var _incoming = flow {
+        while (this@DefaultClientSSESession.coroutineContext.isActive) {
+            while (this@DefaultClientSSESession.coroutineContext.isActive) {
+                val event = input.parseEvent() ?: break
 
-            if (event.isCommentsEvent() && !showCommentEvents) continue
-            if (event.isRetryEvent() && !showRetryEvents) continue
+                if (event.isCommentsEvent() && !showCommentEvents) continue
+                if (event.isRetryEvent() && !showRetryEvents) continue
 
-            send(event)
+                emit(event)
+            }
+
+            if (needToReconnect) {
+                doReconnection()
+            } else {
+                close()
+            }
+        }
+    }
+
+    init {
+        coroutineContext.job.invokeOnCompletion {
+            close()
+        }
+    }
+
+    private suspend fun doReconnection() {
+        var retries = 1
+        while (retries <= maxRetries) {
+            try {
+                withContext(coroutineContext) {
+                    input.cancel()
+
+                    delay(reconnectionTimeMillis)
+
+                    val reconnectionRequest = getRequestForReconnection()
+                    LOGGER.trace(
+                        "Sending SSE request ${reconnectionRequest.url} (attempt ${retries + 1}/${maxRetries + 1})"
+                    )
+
+                    val reconnectionResponse = clientForReconnection.execute(reconnectionRequest).response
+                    LOGGER.trace("Receive response for reconnection SSE request to ${reconnectionRequest.url}")
+                    checkResponse(reconnectionResponse)
+
+                    if (reconnectionResponse.status == HttpStatusCode.NoContent) {
+                        needToReconnect = false
+                    }
+
+                    input = reconnectionResponse.rawContent
+                }
+                return
+            } catch (cause: CancellationException) {
+                close()
+                return
+            } catch (cause: Throwable) {
+                if (retries == maxRetries) {
+                    LOGGER.trace("Max retries ($maxRetries) reached for SSE reconnection, closing session")
+                    close()
+                    throw cause
+                }
+                LOGGER.trace("SSE reconnection attempt ${retries + 1} failed")
+                retries++
```

Comment:
Should we consider some delay policy here?

Reply:
Before every reconnection attempt, we have a [delay](https://github.com/ktorio/ktor/blob/e25f321c017b253673a6cf9d91794070383f295f/ktor-client/ktor-client-core/common/src/io/ktor/client/plugins/sse/DefaultClientSSESession.kt#L72)